File: app/src/main/java/com/example/bindechexclock/MainActivity.java
package com.example.bindechexclock;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import androidx.appcompat.app.AppCompatActivity; // Changed
import com.google.android.material.bottomnavigation.BottomNavigationView; // Changed

import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.ui.AppBarConfiguration;
import androidx.navigation.ui.NavigationUI;

public class MainActivity extends AppCompatActivity {

    TimeManager timeManager = new TimeManager(); // Changed: Removed 'this'

    private Handler handler;
    private Runnable updateTimeRunnable;
    private static final long UPDATE_INTERVAL_MS = 1000; // 1 second

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        BottomNavigationView navView = findViewById(R.id.nav_view);
        // Passing each menu ID as a set of Ids because each
        // menu should be considered as top level destinations.
        AppBarConfiguration appBarConfiguration = new AppBarConfiguration.Builder(
                R.id.navigation_home, R.id.navigation_dashboard, R.id.navigation_notifications,
                R.id.navigation_romans, R.id.navigation_stopwatch)
                .build();
        final NavController navController = Navigation.findNavController(this, R.id.nav_host_fragment);
        NavigationUI.setupActionBarWithNavController(this, navController, appBarConfiguration);
        NavigationUI.setupWithNavController(navView, navController);

        handler = new Handler(Looper.getMainLooper());
        updateTimeRunnable = new Runnable() {
            @Override
            public void run() {
                timeManager.updateTime();
                handler.postDelayed(this, UPDATE_INTERVAL_MS);
            }
        };
    }

    @Override
    protected void onStart() {
        super.onStart();
        // Start the periodic updates when the activity becomes visible
        handler.post(updateTimeRunnable);
    }

    @Override
    protected void onStop() {
        super.onStop();
        // Stop the periodic updates when the activity is no longer visible
        handler.removeCallbacks(updateTimeRunnable);
    }

    public TimeManager getTimeManager() {
        return timeManager;
    }
}

File: app/src/main/java/com/example/bindechexclock/MainActivity.kt
package com.example.bindechexclock

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import androidx.appcompat.app.AppCompatActivity
import androidx.navigation.findNavController
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.setupActionBarWithNavController
import androidx.navigation.ui.setupWithNavController
import com.example.bindechexclock.databinding.ActivityMainBinding // Import view binding class

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    val timeManager = TimeManager() // Public property, accessible by fragments

    private lateinit var handler: Handler
    private lateinit var updateTimeRunnable: Runnable

    companion object {
        private const val UPDATE_INTERVAL_MS = 1000L
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val navView = binding.navView // Access BottomNavigationView via binding

        val appBarConfiguration = AppBarConfiguration(
            setOf(
                R.id.navigation_home, R.id.navigation_dashboard, R.id.navigation_notifications,
                R.id.navigation_romans, R.id.navigation_stopwatch
            )
        )
        val navController = findNavController(R.id.nav_host_fragment)
        setupActionBarWithNavController(navController, appBarConfiguration)
        navView.setupWithNavController(navController)

        handler = Handler(Looper.getMainLooper())
        updateTimeRunnable = object : Runnable {
            override fun run() {
                timeManager.updateTime()
                handler.postDelayed(this, UPDATE_INTERVAL_MS)
            }
        }
    }

    override fun onStart() {
        super.onStart()
        // Start the periodic updates when the activity becomes visible
        handler.post(updateTimeRunnable)
    }

    override fun onStop() {
        super.onStop()
        // Stop the periodic updates when the activity is no longer visible
        handler.removeCallbacks(updateTimeRunnable)
    }

    // No need for getTimeManager() method, fragments can access the public timeManager property.
}
